builtin/common/misc_helpers.lua

```diff
@@ -517,6 +517,66 @@ function table.shuffle(t, from, to, random)
 end
 
 
+function table.find(t, e, i, j)
+	i = i or 1
+	j = j or #t
+    
+    	local mt = {__eq = function(e1, e2)
+     		if type(e2) ~= "table" then
+                	return false
+            	end
+             
+            	local is_elem_exists = true
+            	for i1, v1 in pairs(e1) do
+                	if v1 ~= e2[i1] then
+                    		is_elem_exists = false
+                    		break
+                	end  
+            	end
+                    
+            	return is_elem_exists
+    	end}
+  
+    	if type(e) == "table" then
+        	e = table.copy(e)
+        	setmetatable(e, mt)
+    	end
+    
+    	local iter = 0
+    	local findings = {}
+    	for k, v in pairs(t) do
+        	iter = iter + 1
+        
+        	if type(v) == "table" then
+           		v = table.copy(v)
+            		setmetatable(v, mt)
+        	end
+        	if not(iter < i or iter > j) and e == v then
+            		table.insert(findings, k)
+        	end
+    	end
+    
+    	return findings
+end
+
+
+function table.find_if(t, func, i, j)
+    	i = i or 1
+    	j = j or #t
```

Comment:
The `#t` mnemonic doesn't work if the table is an associative array. I think it returns zero. If this function is truly available only for lists, then `pairs` below should be changed to `ipairs`. Otherwise the range parameters should be removed.

Reply:
Really? Then do I need to write an own a len function that would work with the associative arrays?

Reply:
I have just tested and that operator really works only on integer keys in a table.